Add a script to check product release versions (#62)
mongodb/code-example-tooling: github-check-releases/README.md

# GitHub Check Releases

This directory contains tooling to enable us to check GitHub releases programmatically, and validate them against the
version(s) of the product(s) we use in our code example test suite(s).

This is a simple PoC with three main components:

- A [JSON file](repo-details.json) that contains the list of products, GitHub repo information, and test suite versions.
- It uses [octokit](https://github.com/octokit/octokit.js) to get release data for those MongoDB products from GitHub, and compare it against the version 
  in our JSON file.
- It uses [axios](https://github.com/axios/axios) with the Jira API to automatically create tickets to update test suite 
  versions.

## Add or change products and test suite versions

This project pulls data from [repo-details.json](repo-details.json) to check the release versions for the MongoDB

> products we use in our code example test suites. It deserializes the data using `RepoDetails` declared in the
> [RepoDetails file](RepoDetails.js) to map the owner and repo name for a given product whose version we want to check.

Reviewer: ```suggestion
products we use in our code example test suites.
```
this seems in-the-weeds for a README. maybe put this in the class file as a comment?

### Add a new product

> To add a new product, create a new entry for it in the `repo-details.json` file. 

Reviewer: ```suggestion
To add a new product, create a new entry in the `repo-details.json` file in the following format:

`
    "owner": "<repo-owner>",
    "repo": "<repo-name>",
    "productName": "<MongoDB Product Name>",
    "testSuiteVersion": "<x.y.z>"
`
  
```
(this prob won't format correctly because of the suggestion syntax, but you get the idea)

You can get the owner and name from the repo URL: `https://github.com/<owner>/<repo>`
For example, to add `https://github.com/mongodb/node-mongodb-native`, set `mongodb` as the

> owner and `node-mongodb-native` as the repo name.
> 
> You can get the test suite version from the test suite's dependency file. For example, the Node.js Driver version is in
> `code-example-tests/javascript/driver/package.json`, in the `dependencies` stanza:
> 
> ```
> "dependencies": {
>   "bluehawk": "^1.6.0",
>   "mongodb": "6.17.0" // This is the test suite version for the Driver
> }
> ```
> 
> The product name is just a human-readable name we can use in the output to report on the changes - i.e. `"Node.js Driver"`.
> 
> ### Change the test suite version
> 
> If the version is out of date, bump the version in the test suite, and then update it in [repo-details.json](repo-details.json) here
> in this project.

Reviewer: ```suggestion
For example, if you want to add the Node Driver, you'd add the following to the `repo-details.json`: 

`
"owner": "mongodb", // repo owner from URL
"repo": "node-mongodb-native", // repo name from URL
"productName": "Node.js Driver", // unique, human-readable product name to use in report outputs
"testSuiteVersion": "v6.17.0" // taken from the `package.json` dependencies
`

- repo URL: `https://github.com/mongodb/node-mongodb-native` 
- product name: the "Node.js Driver" 
- the test suite's dependency file's 
`code-example-tests/javascript/driver/package.json` `dependencies` stanza:

`
"dependencies": {
  "bluehawk": "^1.6.0",
  "mongodb": "6.17.0" // the test suite version for the Driver
}
`

### Change the test suite version

> IMPORTANT: You must manually update the version in this project and in the test suite.

If the version is out of date, bump the version in the test suite, and then update it in [repo-details.json](repo-details.json) 
in this project.
```
(again, formatting is slightly wrong to keep the suggestion from ending early)

## GitHub release information

### Get release information from GitHub

This is a simple PoC that uses [octokit](https://github.com/octokit/octokit.js) to get release data for MongoDB products from GitHub.

> This code is in the `check-latest-releases.js` file.

Reviewer: ```suggestion
This code is in the [`check-latest-releases.js`](check-latest-releases.js) file.
```

## Run the tool

### Prerequisites

To run the tool, you need:

**GitHub**:

- A [GitHub Personal Access Token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens) (PAT) with `repo` permissions

> For this project, as a MongoDB org member, you must also auth your PAT with SSO.

Reviewer: ```suggestion
You must also authorize your PAT with SSO for this project as a MongoDB org member.
```

**Jira**:

- A Jira PAT - refer to the internal wiki for instructions on how to create it
- Information about the Jira base URL, project, and component(s) to create the ticket with the appropriate data

**System**:

- Node.js/npm installed

### Steps

1. **Create a `.env` file**

   Create a `.env` file that contains the following details:

   ```
   GITHUB_TOKEN="yourToken"
   JIRA_BASE_URL="https://your-jira-instance-url"
   JIRA_API_TOKEN="your-api-token"
   JIRA_PROJECT_KEY="your-project-key"  
   JIRA_ISSUE_TYPE="Task"
   JIRA_COMPONENT="your-team-or-project-component"
   ```

   Replace the placeholder values with your GitHub token, Jira token, and Jira project/component details.

>    > Note: The `.env` file is in the `.gitignore`, so no worries about accidentally committing credentials.

Reviewer: ```suggestion
   > Note: The `.env` file is in the `.gitignore`, but still use caution to avoid accidentally committing credentials.
```
